Python release 10.3.0 -- Please merge Nov 19 AM (#19320)
newrelic/docs-website: src/content/docs/release-notes/agent-release-notes/python-release-notes/python-agent-100300.mdx

---
subject: 'Python agent'
releaseDate: '2024-11-18'
version: 10.3.0
downloadLink: 'https://pypi.python.org/pypi/newrelic'
features: ['Add Valkey support', 'Add TOML configuration file support', 'Add protobuf v5 support for infinite tracing', 'Add support for attaching custom attributes to logs', 'Add support for attaching labels to logs', 'Minor cleanup of HTTPX instrumentation', 'Collect module information periodically']
bugs: []
security: []
---

## Notes

> This release of the Python agent adds support for Valkey, configuration via TOML file, protobuf v5 for infinite tracing, attaching custom attributes and labels to log events, collecting module information periodically, and some minor cleanup of HTTPX instrumentation. Http2 support was validated for [Daphne](https://pypi.org/project/daphne/), [Hypercorn](https://pypi.org/project/Hypercorn/), and [HTTPX](https://pypi.org/project/httpx/).

Reviewer: ```suggestion
This release of the Python agent adds support for [Valkey](https://pypi.org/project/valkey/), configuration via TOML file, protobuf v5 for infinite tracing, attaching custom attributes and labels to log events, collecting module information periodically, and some minor cleanup of HTTPX instrumentation. Http2 support was validated for [Daphne](https://pypi.org/project/daphne/), [Hypercorn](https://pypi.org/project/Hypercorn/), and [HTTPX](https://pypi.org/project/httpx/).
```

The Python agent now allows you to opt-in to adding your custom tags (labels) to agent-forwarded logs. With custom tags on logs, platform engineers can easily filter, search, and correlate log data for faster and more efficient troubleshooting, improved performance, and optimized resource utilization. To learn more about this feature see the [documentation](/docs/logs/logs-context/Custom-tags-agent-forwarder-logs).
 
Install the agent using `easy_install/pip/distribute` via the [Python Package Index](https://pypi.python.org/pypi/newrelic) or download it directly from the [New Relic download site](https://download.newrelic.com/python_agent/release).

## New features

* Add support for Valkey

  * Add instrumentation for [Valkey](https://pypi.org/project/valkey/). Thanks [@jairhenrique](https://github.com/jairhenrique) for the contribution!
  
* Add TOML configuration file support  

  * Support to read TOML configuration is now available. Thanks [@Tatsh](https://github.com/Tatsh) for the contribution!
  

> * Add infinite tracing pb2 file for protobuf v5
> 
>   * This updates the pb2 files used for infinite tracing to v5. 

Reviewer: ```suggestion
* Add infinite tracing pb2 file for protobuf v5

  * This adds an additional pb2 file used for infinite tracing that is compatible with [protobuf](https://pypi.org/project/protobuf/) v5. 
```

* Add support for attaching custom attributes to logs

>   * A new configuration option has been added called `application_logging.forwarding.custom_attributes` that will add custom attributes to all log events when set .

Reviewer: ```suggestion
  * A new configuration option has been added called `application_logging.forwarding.custom_attributes` that will add custom attributes to all log events when set.
```

* Add support for attaching labels to log events

>   * Labels as attributes can now be added to log events.  When `application_logging.forwarding.labels.enabled` is set the values in `labels` will be added to all log events. Labels can be excluded by setting `application_logging.forwarding.labels.exclude`.

Author: ```suggestion
  * Labels as attributes can now be added to log events. When `application_logging.forwarding.labels.enabled` is set the values in `labels` will be added to all log events. Labels can be excluded by setting `application_logging.forwarding.labels.exclude`.
```

>   * Labels as attributes can now be added to log events.  When `application_logging.forwarding.labels.enabled` is set the values in `labels` will be added to all log events. Labels can be excluded by setting `application_logging.forwarding.labels.exclude`.

Reviewer: ```suggestion
  * Labels as attributes can now be added to log events. When `application_logging.forwarding.labels.enabled` is set the values in `labels` will be added to all log events. Labels can be excluded by setting `application_logging.forwarding.labels.exclude`.
```

## Enhancements

* Minor cleanup of HTTPX instrumentation

>   * Some minor cleanup was done of the HTTPX instrumentation.

Reviewer: ```suggestion
  * Some minor cleanup was done in the HTTPX instrumentation.
```

* Validate HTTP/2 support

  * Added tests to ensure HTTP/2 support for the following frameworks:
    * [Daphne](https://pypi.org/project/daphne/)
    * [Hypercorn](https://pypi.org/project/Hypercorn/)
    * [HTTPX](https://pypi.org/project/httpx/)
    
## Bug fixes

* Collect module information from application periodically instead of at startup

>   * Previously, the New Relic Python agent collected the application's module information all at once during startup.  In some extreme cases this would cause a large spike in CPU usage. This enhancement allows the agent to do this incrementally throughout the application's operation, preventing potential memory spikes during agent startup.

Author: ```suggestion
  * Previously, the New Relic Python agent collected the application's module information all at once during startup. In some extreme cases this would cause a large spike in CPU usage. This enhancement allows the agent to do this incrementally throughout the application's operation, preventing potential memory spikes during agent startup.
```
